File: vsd_fleet_ms/vsd_fleet_ms/page/fleet_finance_dashboard/fleet_finance_dashboard.js
// Copyright (c) 2026, VV SYSTEMS DEVELOPER LTD and contributors
// For license information, please see license.txt

frappe.pages["fleet-finance-dashboard"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Fleet Finance"),
		single_column: true,
	});

	const dash = new FleetFinanceDashboard(page);
	dash.init();
};

class FleetFinanceDashboard {
	constructor(page) {
		this.page = page;
		this.from_date = frappe.datetime.month_start();
		this.to_date   = frappe.datetime.get_today();
	}

	init() {
		this._add_filter_bar();
		this._render_skeleton();
		this.refresh();
	}

	// ── date filter bar ────────────────────────────────────────────────────────

	_add_filter_bar() {
		const bar = $(`
			<div class="fleet-filter-bar">
				<button class="btn btn-sm btn-default fleet-date-btn active" data-range="month">${__("This Month")}</button>
				<button class="btn btn-sm btn-default fleet-date-btn" data-range="last_month">${__("Last Month")}</button>
				<button class="btn btn-sm btn-default fleet-date-btn" data-range="quarter">${__("This Quarter")}</button>
				<button class="btn btn-sm btn-default fleet-date-btn" data-range="year">${__("This Year")}</button>
				<span style="margin:0 8px;color:#aaa">|</span>
				<input type="date" class="fleet-date-input" id="fin-from" value="${this.from_date}">
				<span style="margin:0 4px">—</span>
				<input type="date" class="fleet-date-input" id="fin-to" value="${this.to_date}">
				<button class="btn btn-sm btn-primary" id="fin-apply">${__("Apply")}</button>
			</div>
		`).appendTo(this.page.main);

		bar.on("click", ".fleet-date-btn", (e) => {
			bar.find(".fleet-date-btn").removeClass("active");
			$(e.currentTarget).addClass("active");
			const range = $(e.currentTarget).data("range");
			const dates = this._date_range(range);
			this.from_date = dates[0];
			this.to_date   = dates[1];
			$("#fin-from").val(this.from_date);
			$("#fin-to").val(this.to_date);
			this.refresh();
		});

		$("#fin-apply", bar).on("click", () => {
			this.from_date = $("#fin-from").val();
			this.to_date   = $("#fin-to").val();
			bar.find(".fleet-date-btn").removeClass("active");
			this.refresh();
		});
	}

	_date_range(range) {
		const today = frappe.datetime.get_today();
		if (range === "month") return [frappe.datetime.month_start(), today];
		if (range === "last_month") {
			const d = new Date(today);
			d.setDate(1); d.setMonth(d.getMonth() - 1);
			const s = frappe.datetime.obj_to_str(d);
			const last = new Date(new Date(frappe.datetime.month_start()) - 1);
			return [s, frappe.datetime.obj_to_str(last)];
		}
		if (range === "quarter") {
			const m = new Date().getMonth();
			const q = Math.floor(m / 3);
			const qs = new Date(new Date().getFullYear(), q * 3, 1);
			return [frappe.datetime.obj_to_str(qs), today];
		}
		if (range === "year") return [today.slice(0, 4) + "-01-01", today];
		return [frappe.datetime.month_start(), today];
	}

	// ── skeleton ───────────────────────────────────────────────────────────────

	_render_skeleton() {
		$(this.page.main).append(`
			<div id="fin-dashboard">
				<div class="fleet-section-title">${__("Cash Position")}</div>
				<div id="fin-kpi-row0" class="fleet-kpi-row"></div>

				<div class="fleet-section-title">${__("Revenue & Receivables")}</div>
				<div id="fin-kpi-row1" class="fleet-kpi-row"></div>

				<div class="fleet-section-title">${__("Purchases & Payables")}</div>
				<div id="fin-kpi-row2" class="fleet-kpi-row"></div>

				<div class="fleet-section-title">${__("Profitability")}</div>
				<div id="fin-kpi-row3" class="fleet-kpi-row"></div>

				<div class="fleet-charts-grid">
					<div class="fleet-chart-card fleet-chart-wide">
						<div class="fleet-card-header">${__("Revenue vs Expenses (Monthly)")}</div>
						<div id="fin-monthly-chart"></div>
					</div>
					<div class="fleet-chart-card fleet-chart-wide">
						<div class="fleet-card-header">${__("Cash Flow — Payments In vs Out (Monthly)")}</div>
						<div id="fin-cashflow-chart"></div>
					</div>
					<div class="fleet-chart-card">
						<div class="fleet-card-header">${__("AR Aging (Overdue Receivables)")}</div>
						<div id="fin-aging-chart"></div>
					</div>
					<div class="fleet-chart-card">
						<div class="fleet-card-header">${__("Trip Expense by Type")}</div>
						<div id="fin-expense-type-chart"></div>
					</div>
					<div class="fleet-chart-card">
						<div class="fleet-card-header">${__("Sales Payment Status")}</div>
						<div id="fin-si-status-chart"></div>
					</div>
					<div class="fleet-chart-card">
						<div class="fleet-card-header">${__("Purchase Payment Status")}</div>
						<div id="fin-pi-status-chart"></div>
					</div>
					<div class="fleet-chart-card">
						<div class="fleet-card-header">${__("Top 10 Customers")}</div>
						<div id="fin-customers-chart"></div>
					</div>
					<div class="fleet-chart-card">
						<div class="fleet-card-header">${__("Top 10 Suppliers")}</div>
						<div id="fin-suppliers-chart"></div>
					</div>
				</div>

				<div class="fleet-two-col">
					<div>
						<div class="fleet-section-title">${__("Unpaid Sales Invoices")}</div>
						<div id="fin-si-table"></div>
					</div>
					<div>
						<div class="fleet-section-title">${__("Unpaid Purchase Invoices")}</div>
						<div id="fin-pi-table"></div>
					</div>
				</div>
			</div>
		`);
	}

	// ── main refresh ───────────────────────────────────────────────────────────

	refresh() {
		const M    = "vsd_fleet_ms.vsd_fleet_ms.page.fleet_finance_dashboard.fleet_finance_dashboard";
		const args = { from_date: this.from_date, to_date: this.to_date };
		Promise.all([
			frappe.xcall(`${M}.get_finance_kpis`, args),
			frappe.xcall(`${M}.get_revenue_vs_expense_monthly`, args),
			frappe.xcall(`${M}.get_top_customers`, {...args, limit: 10}),
			frappe.xcall(`${M}.get_top_suppliers`, {...args, limit: 10}),
			frappe.xcall(`${M}.get_unpaid_invoices`, {doctype: "Sales Invoice", limit: 20}),
			frappe.xcall(`${M}.get_unpaid_invoices`, {doctype: "Purchase Invoice", limit: 20}),
			frappe.xcall(`${M}.get_payment_status_breakdown`, args),
			frappe.xcall(`${M}.get_expense_by_type`, args),
			frappe.xcall(`${M}.get_cash_flow_monthly`, args),
			frappe.xcall(`${M}.get_ar_aging`),
		]).then(([kpis, monthly, customers, suppliers, si_unpaid, pi_unpaid, payment_status, exp_types, cashflow, aging]) => {
			this._render_cash_position_kpis(kpis);
			this._render_revenue_kpis(kpis);
			this._render_purchase_kpis(kpis);
			this._render_profit_kpis(kpis);
			this._render_monthly_chart(monthly);
			this._render_cashflow_chart(cashflow);
			this._render_aging_chart(aging);
			this._render_payment_status_charts(payment_status);
			this._render_expense_type_chart(exp_types);
			this._render_customers_chart(customers);
			this._render_suppliers_chart(suppliers);
			this._render_unpaid_table("#fin-si-table", si_unpaid, "Sales Invoice");
			this._render_unpaid_table("#fin-pi-table", pi_unpaid, "Purchase Invoice");
		});
	}

	// ── KPI rows ───────────────────────────────────────────────────────────────

	_render_cash_position_kpis(d) {
		const row = $("#fin-kpi-row0").empty();
		[
			{label: __("Cash Balance"),  value: format_currency(d.cash_balance),  color: d.cash_balance >= 0 ? "blue" : "red",    icon: "🏦", hint: __("All cash & bank accounts")},
			{label: __("Payments In"),   value: format_currency(d.payments_in),   color: "green",  icon: "⬇️", hint: __("Period received")},
			{label: __("Payments Out"),  value: format_currency(d.payments_out),  color: "orange", icon: "⬆️", hint: __("Period paid out")},
			{label: __("Net Cash Flow"), value: format_currency(d.net_cash),      color: d.net_cash >= 0 ? "green" : "red", icon: "💱"},
		].forEach(i => row.append(this._kpi_card(i)));
	}

	_render_revenue_kpis(d) {
		const row = $("#fin-kpi-row1").empty();
		[
			{label: __("Revenue"),      value: format_currency(d.revenue),            color: "blue",   icon: "💰", hint: d.si_count + " " + __("invoice(s)")},
			{label: __("Collected"),    value: format_currency(d.collected),          color: "green",  icon: "✅", hint: d.collection_rate + "% " + __("collected")},
			{label: __("Outstanding"),  value: format_currency(d.receivables),        color: d.receivables > 0 ? "orange" : "green", icon: "⏳"},
			{label: __("Overdue >30d"), value: format_currency(d.overdue_si_amount), color: d.overdue_si_count > 0 ? "red" : "green", icon: "⚠️", hint: d.overdue_si_count + " " + __("invoice(s)")},
		].forEach(i => row.append(this._kpi_card(i)));
	}

	_render_purchase_kpis(d) {
		const row = $("#fin-kpi-row2").empty();
		[
			{label: __("Purchases"),    value: format_currency(d.purchases),          color: "blue",   icon: "🛒", hint: d.pi_count + " " + __("invoice(s)")},
			{label: __("Paid"),         value: format_currency(d.paid),               color: "green",  icon: "✅", hint: d.payment_rate + "% " + __("paid")},
			{label: __("Payables"),     value: format_currency(d.payables),           color: d.payables > 0 ? "orange" : "green", icon: "⏳"},
			{label: __("Overdue >30d"), value: format_currency(d.overdue_pi_amount), color: d.overdue_pi_count > 0 ? "red" : "green", icon: "⚠️", hint: d.overdue_pi_count + " " + __("invoice(s)")},
		].forEach(i => row.append(this._kpi_card(i)));
	}

	_render_profit_kpis(d) {
		const row = $("#fin-kpi-row3").empty();
		[
			{label: __("Gross Profit"),  value: format_currency(d.gross_profit),  color: d.gross_profit >= 0 ? "green" : "red",  icon: "📈"},
			{label: __("Net Profit"),    value: format_currency(d.net_profit),    color: d.net_profit >= 0 ? "green" : "red",    icon: "💹", hint: d.profit_margin + "% " + __("margin")},
			{label: __("Trip Expenses"), value: format_currency(d.trip_expense),  color: "orange", icon: "🚛"},
		].forEach(i => row.append(this._kpi_card(i)));
	}

	_kpi_card({label, value, color, icon, hint}) {
		return `
			<div class="fleet-kpi-card fleet-kpi-${color}">
				<div class="fleet-kpi-icon">${icon}</div>
				<div class="fleet-kpi-value">${value}</div>
				<div class="fleet-kpi-label">${label}</div>
				${hint ? `<div class="fleet-kpi-hint">${hint}</div>` : ""}
			</div>`;
	}

	// ── charts ─────────────────────────────────────────────────────────────────

	_render_monthly_chart(data) {
		if (!data) return;
		const rev = data.revenue || [];
		const exp = data.expense || [];

		// Merge all periods
		const periods = [...new Set([...rev.map(r => r.period), ...exp.map(r => r.period)])].sort();
		if (!periods.length) return;

		const rev_map = Object.fromEntries(rev.map(r => [r.period, r.amount]));
		const exp_map = Object.fromEntries(exp.map(r => [r.period, r.amount]));

		this._frappe_chart("#fin-monthly-chart", {
			data: {
				labels: periods,
				datasets: [
					{ name: __("Revenue"),   values: periods.map(p => flt(rev_map[p] || 0)) },
					{ name: __("Purchases"), values: periods.map(p => flt(exp_map[p] || 0)) },
				],
			},
			type: "bar",
			height: 240,
			colors: ["#4e79a7","#e15759"],
		});
	}

	_render_payment_status_charts(data) {
		if (!data) return;

		const _render = (selector, rows) => {
			if (!rows || !rows.length) return;
			const labels = rows.map(r => r.payment_status || "Unknown");
			const values = rows.map(r => flt(r.amount || 0));
			this._frappe_chart(selector, {
				data: { labels, datasets: [{ values }] },
				type: "donut",
				height: 200,
				colors: ["#dc3545","#fd7e14","#28a745"],
			});
		};

		_render("#fin-si-status-chart", data.sales);
		_render("#fin-pi-status-chart", data.purchases);
	}

	_render_expense_type_chart(rows) {
		if (!rows || !rows.length) return;
		this._frappe_chart("#fin-expense-type-chart", {
			data: {
				labels: rows.map(r => r.expense_type || "Other"),
				datasets: [{ values: rows.map(r => flt(r.amount || 0)) }],
			},
			type: "pie",
			height: 200,
		});
	}

	_render_customers_chart(rows) {
		if (!rows || !rows.length) return;
		this._frappe_chart("#fin-customers-chart", {
			data: {
				labels: rows.map(r => r.customer || "—"),
				datasets: [
					{ name: __("Revenue"),   values: rows.map(r => flt(r.revenue || 0)) },
					{ name: __("Collected"), values: rows.map(r => flt(r.collected || 0)) },
				],
			},
			type: "bar",
			height: 220,
			colors: ["#4e79a7","#59a14f"],
		});
	}

	_render_suppliers_chart(rows) {
		if (!rows || !rows.length) return;
		this._frappe_chart("#fin-suppliers-chart", {
			data: {
				labels: rows.map(r => r.supplier || "—"),
				datasets: [
					{ name: __("Purchases"), values: rows.map(r => flt(r.purchases || 0)) },
					{ name: __("Paid"),      values: rows.map(r => flt(r.paid || 0)) },
				],
			},
			type: "bar",
			height: 220,
			colors: ["#e15759","#59a14f"],
		});
	}

	_render_cashflow_chart(rows) {
		if (!rows || !rows.length) return;
		const periods = rows.map(r => r.period);
		this._frappe_chart("#fin-cashflow-chart", {
			data: {
				labels: periods,
				datasets: [
					{ name: __("Cash In"),  values: rows.map(r => flt(r.cash_in  || 0)) },
					{ name: __("Cash Out"), values: rows.map(r => flt(r.cash_out || 0)) },
				],
			},
			type: "bar",
			height: 240,
			colors: ["#59a14f", "#e15759"],
		});
	}

	_render_aging_chart(data) {
		if (!data) return;
		const buckets = ["0-30", "31-60", "61-90", "90+"];
		const values  = buckets.map(b => flt(data[b] || 0));
		if (!values.some(v => v > 0)) return;
		this._frappe_chart("#fin-aging-chart", {
			data: {
				labels: buckets.map(b => b + " " + __("days")),
				datasets: [{ values }],
			},
			type: "bar",
			height: 200,
			colors: ["#59a14f", "#f28e2b", "#e15759", "#b07aa1"],
		});
	}

	_frappe_chart(selector, config) {
		const el = $(selector)[0];
		if (!el) return;
		try {
			new frappe.Chart(el, {
				...config,
				tooltipOptions: { formatTooltipX: d => d },
			});
		} catch(e) {
			console.warn("Chart error:", e);
		}
	}

	// ── unpaid invoice tables ──────────────────────────────────────────────────

	_render_unpaid_table(selector, rows, doctype) {
		const el = $(selector).empty();
		if (!rows || !rows.length) {
			el.html('<p class="text-muted">' + __("All invoices paid.") + '</p>');
			return;
		}

		const slug = doctype === "Sales Invoice" ? "sales-invoice" : "purchase-invoice";
		let html = `<table class="fleet-table">
			<thead><tr>
				<th>${__("Invoice")}</th><th>${__("Date")}</th>
				<th>${__("Party")}</th><th>${__("Total")}</th>
				<th>${__("Outstanding")}</th><th>${__("Status")}</th>
			</tr></thead><tbody>`;

		rows.forEach(r => {
			const days = Math.round((new Date() - new Date(r.posting_date)) / 86400000);
			const age_class = days > 60 ? "red" : days > 30 ? "orange" : "";
			html += `<tr>
				<td><a href="/app/${slug}/${r.name}">${r.name}</a></td>
				<td class="${age_class}">${r.posting_date}</td>
				<td>${r.party || "—"}</td>
				<td>${format_currency(r.grand_total)}</td>
				<td style="color:red;font-weight:bold">${format_currency(r.outstanding_amount)}</td>
				<td>${r.payment_status || "Unpaid"}</td>
			</tr>`;
		});

		el.html(html + "</tbody></table>");
	}
}

function flt(v) { return parseFloat(v || 0); }

function format_currency(v) {
	return parseFloat(v || 0).toLocaleString(undefined, {
		minimumFractionDigits: 0,
		maximumFractionDigits: 0,
	});
}

File: vsd_fleet_ms/vsd_fleet_ms/page/fleet_finance_dashboard/fleet_finance_dashboard.py
# Copyright (c) 2026, VV SYSTEMS DEVELOPER LTD and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt, nowdate


@frappe.whitelist()
def get_finance_kpis(from_date=None, to_date=None):
    from_date = from_date or nowdate()[:7] + "-01"
    to_date   = to_date   or nowdate()
    params    = {"from_date": from_date, "to_date": to_date}

    # Sales
    si = frappe.db.sql("""
        SELECT
            COUNT(*)                        AS invoice_count,
            IFNULL(SUM(grand_total), 0)     AS revenue,
            IFNULL(SUM(paid_amount), 0)     AS collected,
            IFNULL(SUM(outstanding_amount), 0) AS receivables
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
    """, params, as_dict=True)[0]

    # Purchases
    pi = frappe.db.sql("""
        SELECT
            COUNT(*)                        AS invoice_count,
            IFNULL(SUM(grand_total), 0)     AS purchases,
            IFNULL(SUM(paid_amount), 0)     AS paid,
            IFNULL(SUM(outstanding_amount), 0) AS payables
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
    """, params, as_dict=True)[0]

    # Ledger entries
    le = frappe.db.sql("""
        SELECT
            IFNULL(SUM(CASE WHEN entry_type='Income'  THEN amount ELSE 0 END), 0) AS income,
            IFNULL(SUM(CASE WHEN entry_type='Expense' THEN amount ELSE 0 END), 0) AS expense
        FROM `tabLedger Entry`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
    """, params, as_dict=True)[0]

    # Payments
    payments = frappe.db.sql("""
        SELECT
            IFNULL(SUM(CASE WHEN payment_type='Receive' THEN paid_amount ELSE 0 END), 0) AS received,
            IFNULL(SUM(CASE WHEN payment_type='Pay'     THEN paid_amount ELSE 0 END), 0) AS paid
        FROM `tabPayment Entry`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
    """, params, as_dict=True)[0]

    # Trip expenses (approved fund rows)
    trip_exp = frappe.db.sql("""
        SELECT IFNULL(SUM(fd.request_amount), 0) AS trip_expense
        FROM `tabRequested Fund Details` fd
        INNER JOIN `tabTrips` t ON t.name = fd.parent
        WHERE fd.parenttype = 'Trips'
          AND fd.request_status IN ('Approved', 'Accounts Approved')
          AND t.date BETWEEN %(from_date)s AND %(to_date)s
    """, params, as_dict=True)[0]

    # Overdue receivables > 30 days
    overdue_si = frappe.db.sql("""
        SELECT COUNT(*) AS cnt, IFNULL(SUM(outstanding_amount), 0) AS amount
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND outstanding_amount > 0
          AND posting_date < DATE_SUB(CURDATE(), INTERVAL 30 DAY)
    """, as_dict=True)[0]

    # Overdue payables > 30 days
    overdue_pi = frappe.db.sql("""
        SELECT COUNT(*) AS cnt, IFNULL(SUM(outstanding_amount), 0) AS amount
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1 AND outstanding_amount > 0
          AND posting_date < DATE_SUB(CURDATE(), INTERVAL 30 DAY)
    """, as_dict=True)[0]

    # Actual cash / bank balance from GL
    cash_balance_row = frappe.db.sql("""
        SELECT IFNULL(SUM(gle.debit - gle.credit), 0) AS balance
        FROM `tabGL Entry` gle
        JOIN `tabAccount` acc ON acc.name = gle.account
        WHERE acc.account_sub_type IN ('Cash', 'Bank') AND acc.is_group = 0
    """, as_dict=True)
    cash_balance = flt(cash_balance_row[0].balance) if cash_balance_row else 0.0

    revenue      = flt(si.revenue)
    purchases    = flt(pi.purchases)
    collected    = flt(si.collected)
    paid_out     = flt(pi.paid)
    gross_profit = revenue - purchases
    net_profit   = gross_profit + flt(le.income) - flt(le.expense)
    profit_margin   = round(net_profit / revenue   * 100, 1) if revenue   > 0 else 0.0
    collection_rate = round(collected  / revenue   * 100, 1) if revenue   > 0 else 0.0
    payment_rate    = round(paid_out   / purchases * 100, 1) if purchases > 0 else 0.0

    return {
        "revenue":           revenue,
        "collected":         collected,
        "receivables":       flt(si.receivables),
        "si_count":          int(si.invoice_count or 0),
        "purchases":         purchases,
        "paid":              paid_out,
        "payables":          flt(pi.payables),
        "pi_count":          int(pi.invoice_count or 0),
        "ledger_income":     flt(le.income),
        "ledger_expense":    flt(le.expense),
        "trip_expense":      flt(trip_exp.trip_expense),
        "gross_profit":      gross_profit,
        "net_profit":        net_profit,
        "profit_margin":     profit_margin,
        "collection_rate":   collection_rate,
        "payment_rate":      payment_rate,
        "payments_in":       flt(payments.received),
        "payments_out":      flt(payments.paid),
        "net_cash":          flt(payments.received) - flt(payments.paid),
        "cash_balance":      cash_balance,
        "overdue_si_count":  int(overdue_si.cnt or 0),
        "overdue_si_amount": flt(overdue_si.amount),
        "overdue_pi_count":  int(overdue_pi.cnt or 0),
        "overdue_pi_amount": flt(overdue_pi.amount),
    }


@frappe.whitelist()
def get_revenue_vs_expense_monthly(from_date=None, to_date=None):
    from_date = from_date or (nowdate()[:4] + "-01-01")
    to_date   = to_date   or nowdate()
    params    = {"from_date": from_date, "to_date": to_date}

    revenue_rows = frappe.db.sql("""
        SELECT DATE_FORMAT(posting_date,'%%Y-%%m') AS period,
               IFNULL(SUM(grand_total), 0) AS amount
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY period ORDER BY period
    """, params, as_dict=True)

    expense_rows = frappe.db.sql("""
        SELECT DATE_FORMAT(posting_date,'%%Y-%%m') AS period,
               IFNULL(SUM(grand_total), 0) AS amount
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY period ORDER BY period
    """, params, as_dict=True)

    return {
        "revenue": revenue_rows,
        "expense": expense_rows,
    }


@frappe.whitelist()
def get_top_customers(from_date=None, to_date=None, limit=10):
    from_date = from_date or nowdate()[:7] + "-01"
    to_date   = to_date   or nowdate()

    rows = frappe.db.sql("""
        SELECT
            customer,
            COUNT(*) AS invoices,
            IFNULL(SUM(grand_total), 0)        AS revenue,
            IFNULL(SUM(paid_amount), 0)        AS collected,
            IFNULL(SUM(outstanding_amount), 0) AS outstanding
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY customer
        ORDER BY revenue DESC
        LIMIT %(limit)s
    """, {"from_date": from_date, "to_date": to_date, "limit": int(limit)}, as_dict=True)

    return rows


@frappe.whitelist()
def get_top_suppliers(from_date=None, to_date=None, limit=10):
    from_date = from_date or nowdate()[:7] + "-01"
    to_date   = to_date   or nowdate()

    rows = frappe.db.sql("""
        SELECT
            supplier,
            COUNT(*) AS invoices,
            IFNULL(SUM(grand_total), 0)        AS purchases,
            IFNULL(SUM(paid_amount), 0)        AS paid,
            IFNULL(SUM(outstanding_amount), 0) AS outstanding
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY supplier
        ORDER BY purchases DESC
        LIMIT %(limit)s
    """, {"from_date": from_date, "to_date": to_date, "limit": int(limit)}, as_dict=True)

    return rows


@frappe.whitelist()
def get_unpaid_invoices(doctype="Sales Invoice", limit=20):
    if doctype not in ("Sales Invoice", "Purchase Invoice"):
        frappe.throw("Invalid doctype")

    if doctype == "Sales Invoice":
        rows = frappe.db.sql("""
            SELECT name, posting_date, customer AS party,
                   grand_total, outstanding_amount, payment_status
            FROM `tabSales Invoice`
            WHERE docstatus = 1 AND outstanding_amount > 0
            ORDER BY posting_date ASC
            LIMIT %(limit)s
        """, {"limit": int(limit)}, as_dict=True)
    else:
        rows = frappe.db.sql("""
            SELECT name, posting_date, supplier AS party,
                   grand_total, outstanding_amount, payment_status
            FROM `tabPurchase Invoice`
            WHERE docstatus = 1 AND outstanding_amount > 0
            ORDER BY posting_date ASC
            LIMIT %(limit)s
        """, {"limit": int(limit)}, as_dict=True)

    return rows


@frappe.whitelist()
def get_payment_status_breakdown(from_date=None, to_date=None):
    from_date = from_date or nowdate()[:7] + "-01"
    to_date   = to_date   or nowdate()
    params    = {"from_date": from_date, "to_date": to_date}

    si_rows = frappe.db.sql("""
        SELECT payment_status, COUNT(*) AS cnt, IFNULL(SUM(grand_total), 0) AS amount
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY payment_status
    """, params, as_dict=True)

    pi_rows = frappe.db.sql("""
        SELECT payment_status, COUNT(*) AS cnt, IFNULL(SUM(grand_total), 0) AS amount
        FROM `tabPurchase Invoice`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY payment_status
    """, params, as_dict=True)

    return {"sales": si_rows, "purchases": pi_rows}


@frappe.whitelist()
def get_expense_by_type(from_date=None, to_date=None):
    from_date = from_date or nowdate()[:7] + "-01"
    to_date   = to_date   or nowdate()

    rows = frappe.db.sql("""
        SELECT
            fd.expense_type,
            IFNULL(SUM(fd.request_amount), 0) AS amount
        FROM `tabRequested Fund Details` fd
        INNER JOIN `tabTrips` t ON t.name = fd.parent
        WHERE fd.parenttype = 'Trips'
          AND fd.request_status IN ('Approved', 'Accounts Approved')
          AND t.date BETWEEN %(from_date)s AND %(to_date)s
          AND fd.expense_type IS NOT NULL AND fd.expense_type != ''
        GROUP BY fd.expense_type
        ORDER BY amount DESC
        LIMIT 10
    """, {"from_date": from_date, "to_date": to_date}, as_dict=True)

    return rows


@frappe.whitelist()
def get_cash_flow_monthly(from_date=None, to_date=None):
    """Monthly cash in (payments received) vs cash out (payments made)."""
    from_date = from_date or (nowdate()[:4] + "-01-01")
    to_date   = to_date   or nowdate()
    params    = {"from_date": from_date, "to_date": to_date}

    rows = frappe.db.sql("""
        SELECT
            DATE_FORMAT(posting_date, '%%Y-%%m') AS period,
            IFNULL(SUM(CASE WHEN payment_type = 'Receive' THEN paid_amount ELSE 0 END), 0) AS cash_in,
            IFNULL(SUM(CASE WHEN payment_type = 'Pay'     THEN paid_amount ELSE 0 END), 0) AS cash_out
        FROM `tabPayment Entry`
        WHERE docstatus = 1 AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        GROUP BY period
        ORDER BY period
    """, params, as_dict=True)

    return rows


@frappe.whitelist()
def get_ar_aging():
    """Accounts Receivable aging buckets (all outstanding Sales Invoices)."""
    buckets = frappe.db.sql("""
        SELECT
            IFNULL(SUM(CASE WHEN DATEDIFF(CURDATE(), posting_date) BETWEEN 0  AND 30  THEN outstanding_amount ELSE 0 END), 0) AS days_0_30,
            IFNULL(SUM(CASE WHEN DATEDIFF(CURDATE(), posting_date) BETWEEN 31 AND 60  THEN outstanding_amount ELSE 0 END), 0) AS days_31_60,
            IFNULL(SUM(CASE WHEN DATEDIFF(CURDATE(), posting_date) BETWEEN 61 AND 90  THEN outstanding_amount ELSE 0 END), 0) AS days_61_90,
            IFNULL(SUM(CASE WHEN DATEDIFF(CURDATE(), posting_date)  > 90              THEN outstanding_amount ELSE 0 END), 0) AS days_90_plus
        FROM `tabSales Invoice`
        WHERE docstatus = 1 AND outstanding_amount > 0
    """, as_dict=True)[0]

    return {
        "0-30":   flt(buckets.days_0_30),
        "31-60":  flt(buckets.days_31_60),
        "61-90":  flt(buckets.days_61_90),
        "90+":    flt(buckets.days_90_plus),
    }
